### PR TITLE
Use Rc for LineEntry to reduce memory consumption

### LINKED ISSUE #388: Use Rc for LineEntry to reduce memory consumption
### Description

The `LineEntry` struct has the [FileEntry](https://github.com/dotenv-linter/dotenv-linter/blob/master/src/common/line_entry.rs#L7) field which clones for every `LineEntry` in the [get_line_entries](https://github.com/dotenv-linter/dotenv-linter/blob/master/src/lib.rs#L229) method.

Use the [Rc](https://doc.rust-lang.org/std/rc/index.html) type for `LineEntry` to reduce memory consumption.

### Implementation

**Step 1:**

Add a new method `new` for `LineEntry` which takes `file` as `Rc<FileEntry>`.

**Step 2:**

Use the method from step 1 to create the `LineEntry` struct in all places.

### PR DESCRIPTION
**Pull request by Tom01098**
Closes #388.

#### ✔ Checklist:
- [x] Commit messages have been written in [Conventional Commits](https://www.conventionalcommits.org) format;
- [x] This PR has been added to [CHANGELOG.md](https://github.com/dotenv-linter/dotenv-linter/blob/master/CHANGELOG.md) (at the top of the list);

### CODE REVIEW

**Comment by Tom01098** [comment]
I was wondering how to enforce that `new` was called, currently anyone could just use the struct creation syntax.

I found [this](https://steveklabnik.com/writing/structure-literals-vs-constructors-in-rust) which recommended adding a private field to the struct. clippy warned against this, using [non_exhaustive](https://doc.rust-lang.org/reference/attributes/type_system.html) instead, but I found this only works at the crate level.

So what's most 'idiomatic'? I'm not sure where to go from here (but that's really irrelevant to this PR!).

File: src/common/line_entry.rs [hunk 1]
```diff
@@ -1,17 +1,29 @@
...
     pub is_deleted: bool,
 }
 
 impl LineEntry {
+    pub fn new(number: usize, file: Rc<FileEntry>, raw_string: String, is_deleted: bool) -> Self {
+        LineEntry {
+            number,
+            file,
+            raw_string,
+            is_deleted,
```
Comment (mgrachev) [inline src/common/line_entry.rs:23]: The `is_deleted` field should be `false` by default.

File: src/common/line_entry.rs [hunk 2]
```diff
@@ -1,17 +1,29 @@
...
-    pub file: FileEntry,
+    pub file: Rc<FileEntry>,
     pub raw_string: String,
 
     /// Used in ExtraBlankLineFixer
     pub is_deleted: bool,
 }
 
 impl LineEntry {
+    pub fn new(number: usize, file: Rc<FileEntry>, raw_string: String, is_deleted: bool) -> Self {
```
Comment (mgrachev) [inline src/common/line_entry.rs:18]:
```suggestion
    pub fn new(number: usize, file: Rc<FileEntry>, raw_string: String) -> Self {
```
The `is_deleted` field should be `false` by default.

**Review by mgrachev — COMMENTED** [review]
@Tom01098 👋 Thank you for your contribution 🚀

I have left some comments. Please take a look at them 👀

**Comment by mgrachev** [comment]
> I was wondering how to enforce that `new` was called, currently anyone could just use the struct creation syntax.
> 
> I found [this](https://steveklabnik.com/writing/structure-literals-vs-constructors-in-rust) which recommended adding a private field to the struct. clippy warned against this, using [non_exhaustive](https://doc.rust-lang.org/reference/attributes/type_system.html) instead, but I found this only works at the crate level.
> 
> So what's most 'idiomatic'? I'm not sure where to go from here (but that's really irrelevant to this PR!).

I like more the way where you can create a public `Struct` with a private field:
```rust
pub struct LineEntry {
    pub number: usize,
    pub file: Rc<FileEntry>,
    pub raw_string: String,

    /// Used in ExtraBlankLineFixer
    is_deleted: bool,
}
```

This struct you can't create like this `LineEntry { ... }` - only using the `new` method.

If you want to deal with the private field you can create a method like this:
```rust
pub fn is_deleted(&self) -> bool {
    self.is_deleted
}
```

**Comment by Tom01098** [comment]
@mgrachev thanks for the review, I should have corrected everything as requested. (If only I remembered conventional commits didn't end with a full stop I wouldn't have to keep force pushing!!)

**Review by mgrachev — APPROVED** [review]

**Comment by mgrachev** [comment]
@Tom01098 Thank you for your help ❤️

**Pull request merged**